contrib/format-iceberg/src/test/java/org/apache/drill/exec/store/iceberg/IcebergQueriesTest.java

```diff
@@ -0,0 +1,681 @@
+
+    String query = "select * from table(dfs.tmp.testAllTypes(type => 'iceberg', snapshotId => %s))";
+
+    long snapshotId = queryBuilder().sql(snapshotQuery).singletonLong();
+    String plan = queryBuilder().sql(query, snapshotId).explainJson();
+    long count = queryBuilder().physical(plan).run().recordCount();
+    assertEquals(2, count);
+  }
+
+  @Test
+  public void testSelectWithSnapshotAsOfTime() throws Exception {
+    String snapshotQuery = "select committed_at from dfs.tmp.`testAllTypes#snapshots` order by committed_at limit 1";
+
+    String query = "select * from table(dfs.tmp.testAllTypes(type => 'iceberg', snapshotAsOfTime => %s))";
+
+    long snapshotId = queryBuilder().sql(snapshotQuery).singletonLong();
+    String plan = queryBuilder().sql(query, snapshotId).explainJson();
+    long count = queryBuilder().physical(plan).run().recordCount();
+    assertEquals(2, count);
+  }
+
+  @Test
+  public void testSelectFromSnapshotId() throws Exception {
+    String snapshotQuery = "select snapshot_id from dfs.tmp.`testAllTypes#snapshots` order by committed_at limit 1";
+
+    String query = "select * from table(dfs.tmp.testAllTypes(type => 'iceberg', fromSnapshotId => %s))";
+
+    long snapshotId = queryBuilder().sql(snapshotQuery).singletonLong();
+    String plan = queryBuilder().sql(query, snapshotId).explainJson();
+    long count = queryBuilder().physical(plan).run().recordCount();
+    assertEquals(1, count);
+  }
+
+  @Test
+  public void testSelectFromSnapshotIdAndToSnapshotId() throws Exception {
+    String snapshotQuery = "select snapshot_id from dfs.tmp.`testAllTypes#snapshots` order by committed_at";
+
+    String query = "select * from table(dfs.tmp.testAllTypes(type => 'iceberg', fromSnapshotId => %s, toSnapshotId => %s))";
+
+    DirectRowSet rowSet = queryBuilder().sql(snapshotQuery).rowSet();
+    try {
+      RowSetReader reader = rowSet.reader();
+      assertTrue(reader.next());
+      Long fromSnapshotId = (Long) reader.column(0).reader().getObject();
+      assertTrue(reader.next());
+      Long toSnapshotId = (Long) reader.column(0).reader().getObject();
+
+      String plan = queryBuilder().sql(query, fromSnapshotId, toSnapshotId).explainJson();
+      long count = queryBuilder().physical(plan).run().recordCount();
+      assertEquals(1, count);
+    } finally {
+      rowSet.clear();
+    }
+  }
+
+  @Test
+  public void testSelectWithSnapshotIdAndSnapshotAsOfTime() throws Exception {
+    String query = "select * from table(dfs.tmp.testAllTypes(type => 'iceberg', snapshotId => %s, snapshotAsOfTime => %s))";
+    try {
+      queryBuilder().sql(query, 123, 456).run();
+    } catch (UserRemoteException e) {
+      MatcherAssert.assertThat(e.getVerboseMessage(), containsString("Both 'snapshotId' and 'snapshotAsOfTime' cannot be specified"));
+    }
+  }
+
+  @Test
+  public void testAllTypes() throws Exception {
```

Comment:
Does iceberg support arrays?  Would it be worth including a unit test for that?

Reply:
Yes, it supports arrays and this test also checks that it can be read correctly. Test below also checks that project pushdown for the case of selecting array indexes works correctly.